### PR TITLE
Multiple Output Datasets & Dynamic Partitioning of PartitionedFileSet

### PR DESCRIPTION
**Pull request by anwar6953**
Documentation for:
1) Multiple Output Datasets of a MapReduce program.
2) Dynamic Partitioning of PartitionedFileSet, as output of a MapReduce program.

https://issues.cask.co/browse/CDAP-3687

### CODE REVIEW

File: cdap-docs/developers-manual/source/building-blocks/datasets/datasets-mapreduce.rst [hunk 1]
```diff
@@ -51,7 +51,7 @@ To read a range of keys and give a hint that you want 16 splits, write::
   ...
   public void beforeSubmit(MapReduceContext context) throws Exception {
     ...
-    context.setInput(kvTable, kvTable.getSplits(16, startKey, stopKey);
+    context.setInput("myTable", kvTable.getSplits(16, startKey, stopKey);
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/datasets/datasets-mapreduce.rst]: there seems to be a syntax error here - a missing ")" - maybe better to copy this from the original code ?

File: cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst [hunk 2]
```diff
@@ -318,6 +318,43 @@ For example, give these arguments when starting the MapReduce through a RESTful
     "dataset.totals.output.partition.key.league" : "nfl"
   }
 
+Dynamic Partitioning of MapReduce Output
+========================================
+
+A MapReduce job can also write to multiple partitions of a PartitionedFileSet, using the
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst]: remove "also"?

File: cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst [hunk 3]
```diff
@@ -318,6 +318,43 @@ For example, give these arguments when starting the MapReduce through a RESTful
...
+    }
+  }
+
+Then, set the class of the custom partitioner as runtime arguments of the output PartitionedFileSet::
+
+  Map<String, String> cleanRecordsArgs = new HashMap<>();
+  PartitionedFileSetArguments.setDynamicPartitioner(cleanRecordsArgs, TimeAndZipPartitioner.class);
+  context.addOutput(DataCleansing.CLEAN_RECORDS, cleanRecordsArgs);
+
+Then, for each record processed by the MapReduce job, the record will be written to a path corresponding
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst]: With this, each record processed by the MapReduce job will be written

File: cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst [hunk 4]
```diff
@@ -318,6 +318,43 @@ For example, give these arguments when starting the MapReduce through a RESTful
...
+to the ``Partition`` that it was mapped to by the ``DynamicPartitioner``, and the set of new ``Partition``s
+will be registered with the output ``PartitionedFileSet`` at the end of the job.
+
+
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst]: Can you explain what happens if any of the dynamic partitions already exist?

File: cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst [hunk 5]
```diff
@@ -134,20 +134,8 @@ MapReduce and Datasets
...
-
-     public class MyMapReduceJob implements MapReduce {
-       @Override
-       public void configure(MapReduceConfigurer configurer) {
-         ...
-         useDatasets(Arrays.asList("catalog"))
-         ...
-
-#. Inject the dataset into the mapper or reducer that uses it::
+To access a dataset directly in mapper or reducer,
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: in the mapper

**Comment by anew** [comment]
I had a few minor comments.

**Comment by awholegunch** [comment]
I've made a number edits. Please check. Otherwise, LGTM.

**Comment by anwar6953** [comment]
@awholegunch changes look good to me.

**Comment by anew** [comment]
@anwar6953 please address this comment before merging: https://github.com/caskdata/cdap/pull/4120#discussion_r40041693

**Comment by awholegunch** [comment]
@anew I think @anwar6953 has addressed this at the end of https://github.com/caskdata/cdap/pull/4120/files#diff-df59c34f2094ac6267c42b8b5e49f3be

**Comment by anwar6953** [comment]
@anew 
The partitions are added at the end of the MapReduce job anyways, and it does fail right away, when an existing partition key is found.

**Comment by awholegunch** [comment]
I think our comments crossed: my comment was in reference to your earlier one...and not your new concern.

**Comment by anew** [comment]
@awholegunch I think I also made my first comment without seeing @anwar6953 's last commit. Somehow I could see yours but not his. All good. We are in sync now.

**Comment by anew** [comment]
LGTM

**Pull request merged**